Added support for unauthenticated user and IP checking

xoen opened the pull request:
## What

The proxy can be configured to not require authentication
(`AUTHENTICATION_REQUIRED` environment variable) and/or check the IP
address of the user (`IP_RE` environment variable).


#### `AUTHENTICATION_REQUIRED` environment variable 

When `AUTHENTICATION_REQUIRED=false` we skip the authentication step (e.g.
we don't ask the user to log in using Auth0). 

By default authentication is required.


### `NICKNAME_RE` environment variable

In addition to the authentication the proxy can also authorise the user
has the correct nickname (`NICKNAME_RE` environment variable). This is checked
only if the authentication is required.


### `IP_RANGES` environment variable

By setting the `IP_RANGES` environment variable we can also check the IP address
of the user as an additional layer of security.

This is independent from the authentication/nickname check. This means the
proxy can be configured to not require login but check the IP or to check
a specific set of users is logging from a specific range of IPs.


### Ticket

https://trello.com/c/UVgHMv1C/223-l-allow-ip-based-access-control-to-shiny-apps-for-dom1-quantum-etc-instead-of-or-as-well-as-user-based-access-controls

xoen commented:
Have a look to see if this looks good to you or there is some case I'm missing.
Let me know if you have any questions.

Also, I added semantic versioning for the docker image as we want to move in that directions for all our images. This shouldn't break compatibility with the way it worked before in theory.

andyhd commented on README.md:
```diff
@@ -15,6 +15,8 @@ environment variables:
 | AUTH0_CALLBACK_URL | OAuth callback URL                                      |
 | AUTH0_PASSWORDLESS | If 'true' use [passwordless authentication](https://auth0.com/docs/connections/passwordless). |
 | NICKNAME_RE        | Regular expression determining who is authorised. When not provided the user nickname is not checked. |
+| IP_RE              | Regular expression determining IPs authorised to access. When not provided the user IP is not checked. |
```
Should there be an example or three of valid regular expressions here?

xoen replied:
I thought I did, I'll add some more examples in the README ;)

andyhd commented on routes/index.js:
```diff
@@ -25,12 +27,35 @@ function nicknameFromRequest(req) {
+    const re = new RegExp(env.IP_RE);
+    var ip = ipFromRequest(req)
+
+    if (!re.test(ip)) {
```
What exactly is this test for?
Not sure regular expressions are the best way to check if IP addresses are in a given range. Ranges are normally written in [CIDR notation](https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing#CIDR_notation), but regexes won't handle that.
Also, IP addresses contain period characters (`.`), which will match any single character in a regular expression unless escaped, meaning the `IP_RE` regex will usually have to be escaped. Eg: `IP_RE=1.2.3.4` will match `1x283=4`
Maybe something like [this SO answer](https://stackoverflow.com/a/29173281) would be better?

xoen replied:
I'm really wary of writing something similar to what was in the SO answer but looks like there is a node package doing this already: https://www.npmjs.com/package/cidr-js

I would still have to pass a comma separated list of CIDR IP ranges and build multiple lists from different blocks but still better than handling everything manually.

Also, I'll have to see if this package actually work.

xoen replied:
Regarding the dots in the regular expressions, yes they need escaping as `.` means "any characters" otherwise.

We may have something like `^(127\.0\.0\.[0-16-9])$` for something like `127.0.0.0/31,127.0.0.6/31,127.0.0.8/31` (CIDR) or `127.0.0.0,127.0.0.1,127.0.0.6,127.0.0.7,127.0.0.8,127.0.0.9` (list)

xoen replied:
The [ipcheck](https://github.com/gosquared/ipcheck) node module may be better than `cidr-js` as far I can tell it seems to handle the "no mask" scenario better, in that case it just uses the `/32` mask.

andyhd replied:
👍

andyhd commented on routes/index.js:
```diff
@@ -73,14 +98,20 @@ router.all('/favicon.ico', function(req, res, next) {
-/* Authenticate and proxy all other requests */
-router.all(/.*/, ensureLoggedIn, function(req, res, next) {
-  var nickname = nicknameFromRequest(req);
+function checkRequiresAuthentication(req, res, next) {
```
The name of this function is misleading, because it not only checks whether authentication is required, but also performs authentication. I would suggest renaming it to `handleAuthentication` or even just `authenticate`.

xoen replied:
Well, that middleware check if the proxy requires authentication or not (`AUTHENTICATION_REQUIRED` environment variable). The name of the function is very specific, but maybe `handleAuthentication` would be good enough.

Have another look and let me know if you think it's better to rename.

andyhd replied:
Yes, it is specific, but only for part of what happens when you call it. If it was named explicitly, it would need to be something like `checkRequiresAuthenticationAndAuthenticateIfSo`

xoen replied:
As discussed offline `handleAuthentication` will be good enough, so I'll rename.

xoen commented on circle.yml:
```diff
@@ -3,7 +3,7 @@ machine:
   environment:
     DOCKER_IMAGE_NAME: quay.io/mojanalytics/auth-proxy
-    DOCKER_IMAGE_TAG: $(echo $CIRCLE_SHA1 | cut -c -8)
+    DOCKER_IMAGE_TAG: v0.2.0
```
Yep, we want to move away from hashes for docker tags as you can't tell what is old and what is new, and move to semantic versioning (I think Kerin already started to convert the RStudio docker image)

xoen commented:
Hey @andyhd I switched to CIDR IP ranges instead of using a regular expression.
I also had to handle the case when [`X-Forwarded-For` contains more than one IP](http://docs.aws.amazon.com/en_en/elasticloadbalancing/latest/classic/x-forwarded-headers.html).

(And I renamed the middleware as discussed earlier)

xoen commented:
Also [added semicolons](https://github.com/ministryofjustice/analytics-platform-auth-proxy/pull/2/commits/55443f62376713a5362b6cdc732fe422ea0014db) and [using single quotes](https://github.com/ministryofjustice/analytics-platform-auth-proxy/pull/2/commits/b1ff61ae9a1bff264a01c0945e2c81d69374d4c3) as per Airbnb's JS Style guide/bible.

xoen commented:
Cheers @andyhd 👍
